broker-client/src/test/java/io/camunda/zeebe/broker/client/api/BrokerClientTest.java

```diff
@@ -331,5 +389,99 @@ protected UnifiedRecordValue toResponseDto(final DirectBuffer buffer) {
+    @Test
+    void shouldRouteToLeaderOfRequestedPartition() {
+      // given - a second broker (1), which will respond successfully, and broker 0 which will
+      // respond
+      // with an error, we expect to get a successful response
+      // the request is routed to partition 1 first through the default round robin strategy
+      final var request = new TestCommand();
+      final BrokerResponse<?> response;
+      broker.updateInfo(info -> info.setFollowerForPartition(1));
+      registerError(broker, ErrorCode.INTERNAL_ERROR, "test");
+
+      try (final var otherBroker =
+          new StubBroker(1).start().updateInfo(info -> info.setLeaderForPartition(1, 1))) {
+        registerSuccessResponse(otherBroker);
+
+        topologyManager.event(new ClusterMembershipEvent(Type.METADATA_CHANGED, broker.member()));
+        topologyManager.event(new ClusterMembershipEvent(Type.MEMBER_ADDED, otherBroker.member()));
+        Awaitility.await("Topology is updated")
+            .untilAsserted(
+                () -> assertThat(topologyManager.getTopology().getLeaderForPartition(1)).isOne());
+
+        // when
+        response = client.sendRequest(request).join();
+      }
+
+      // then
+      assertThat(response.isResponse()).isTrue();
+    }
+
+    @Test
+    void shouldRouteRequestBasedOnPartitionId() {
+      // given - a second broker (1), which respond successfully for partition 2
+      final BrokerResponse<?> response;
+      final var request = new TestCommand(1L);
+      request.setPartitionId(2);
+
+      try (final var otherBroker = new StubBroker(1, 2).start()) {
+        registerSuccessResponse(otherBroker);
+        topologyManager.event(new ClusterMembershipEvent(Type.MEMBER_ADDED, otherBroker.member()));
+        Awaitility.await("Topology is updated")
+            .untilAsserted(
+                () -> assertThat(topologyManager.getTopology().getLeaderForPartition(2)).isOne());
+
+        // when
+        response = client.sendRequest(request).join();
+      }
+
+      // then
+      assertThat(response.isResponse()).isTrue();
+    }
+
+    @Test
+    void shouldRouteRequestBasedOnDispatchStrategy() {
+      // given - a second broker (1), which respond successfully for partition 2
+      final BrokerResponse<?> response;
+      final var request = new TestCommand(1L, ignored -> 2);
+
+      try (final var otherBroker = new StubBroker(1, 2).start()) {
+        registerSuccessResponse(otherBroker);
+        topologyManager.event(new ClusterMembershipEvent(Type.MEMBER_ADDED, otherBroker.member()));
+        Awaitility.await("Topology is updated")
+            .untilAsserted(
+                () -> assertThat(topologyManager.getTopology().getLeaderForPartition(2)).isOne());
+
+        // when
+        response = client.sendRequest(request).join();
+      }
+
+      // then
+      assertThat(response.isResponse()).isTrue();
+    }
+
+    @Test
+    void shouldRouteToDeploymentPartitionAsFallback() {
```

Comment:
:question: Should it route to deployment partition as fall back? Shouldn't it fail if dispatch strategy doesn't return a valid partition?

Reply:
:+1: I also don't know if there is a reason for this. It is ok to leave it like this as it is not part of this PR. But would be nice to check it later.